### PR TITLE
feat: update query builder to support spatial aggregations and functions

### PR DESCRIPTION
**Pull request by YounixM**
UI Changes related to - https://github.com/SigNoz/signoz/issues/4016

Screenshot 2024-02-19 005732
image

Impacted Areas:

Dashboard
Alerts

## Summary by CodeRabbit

- **New Features**
	- Introduced support for multiple API versions, enabling queries with API v4.
	- Added new metrics operators for advanced data aggregation and analysis.
	- Implemented keyboard shortcuts for dashboard and query builder functionalities, enhancing user interaction efficiency.
	- Expanded query builder capabilities with new functions for complex data manipulation and analysis.
	- Enhanced dashboard components with new styling options and improved user interface elements.

- **Enhancements**
	- Updated aggregation methods in metrics queries to use average instead of sum for more accurate data representation.
	- Improved tooltip functionality in graphs for better data visualization.

- **Bug Fixes**
	- Addressed an issue with duplicate legend labels in tooltips, ensuring accurate data representation.
	- Fixed a bug in query parameter handling to correctly apply versioning in dashboard and metrics queries.

- **Refactor**
	- Optimized query range parameter parsing for improved performance and accuracy in time-shifting functionalities.
	- Refactored the handling of metric limits to ensure consistent application across queries.

### CODE REVIEW

**Comment by srikanthccv** [comment]
- If the version is not present default version to v3
- With the above fallback in place, whenever the dashboard version is v3, let's continue to show the old UI for metrics builder.
- With some recent commit, the endpoint logic is not working properly. All requests are going to /v3
- For the table type, the column for aggregation should be `<SpaceAggregator>(metric_name)` for v4, instead of the `count(metric_name)`
- Changing the metric name doesn't properly update the time and space aggregation operators
- When there are two lines with the same labels, the legend has duplicate values.

File: frontend/src/api/metrics/getQueryRange.ts [hunk 1]
```diff
@@ -9,10 +9,23 @@ import {
 
 export const getMetricsQueryRange = async (
 	props: QueryRangePayload,
+	version: string | undefined,
```
Comment (Rajat-Dabade) [inline frontend/src/api/metrics/getQueryRange.ts:12]: NIT: `version?: string`

File: frontend/src/container/QueryBuilder/components/QueryFunctions/QueryFunctions.tsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+		setFunctions(clonedFunctions);
+		onChange(clonedFunctions);
+	};
+
+	const handleUpdateFunctionName = (
+		func: QueryFunctionProps,
+		index: number,
+		value: string,
+	): void => {
+		console.log(func);
```
Comment (Rajat-Dabade) [inline frontend/src/container/QueryBuilder/components/QueryFunctions/QueryFunctions.tsx:112]: console.

File: frontend/src/container/QueryBuilder/components/QueryFunctions/QueryFunctions.tsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+			onChange(updateFunctions);
+		}
+	};
+
+	const handleUpdateFunctionArgs = (
+		func: QueryFunctionProps,
+		index: number,
+		value: string,
+	): void => {
+		console.log(func, index, value);
```
Comment (Rajat-Dabade) [inline frontend/src/container/QueryBuilder/components/QueryFunctions/QueryFunctions.tsx:128]: console.

**Comment by YounixM** [comment]
If the version is not present default version to v3 ✅ 
With the above fallback in place, whenever the dashboard version is v3, let's continue to show the old UI for the metrics builder. ✅ 
With some recent commits, the endpoint logic is not working properly. All requests are going to /v3 ✅ 
Changing the metric name doesn't properly update the time and space aggregation operators ✅

**Comment by srikanthccv** [comment]
@YounixM please fix the removed list view from panel types.

**Comment by YounixM** [comment]
Yes @srikanthccv, aware of this issue. I have to verify the file with develop to make sure existing changes are not removed. That work is pending

**Review by srikanthccv — APPROVED** [review]

**Pull request merged**